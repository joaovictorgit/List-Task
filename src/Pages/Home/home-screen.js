import {
  View,
  Text,
  TouchableOpacity,
  Modal,
  SafeAreaView,
  ScrollView,
} from "react-native";

import styles from "./style";
import { useEffect, useState } from "react";
import { TextInput } from "react-native-gesture-handler";
import { db } from "../firebase/config";
import { onValue, push, ref, remove, set } from "firebase/database";
import moment from "moment";
import uuid from "react-native-uuid";

const Home = () => {
  moment.locale("en");

  const [modalVisible, setModalVisible] = useState(false);
  const [allTasks, setAllTasks] = useState([]);
  const [task, setTask] = useState("");
  const [oldTask, setOldTask] = useState();
  const [statusAddTask, setStatusAddTask] = useState(true);

  const addTask = () => {
    if (statusAddTask) {
      const id = uuid.v4();
      set(ref(db, "tarefas/" + id), {
        id: id,
        task: task,
        statusCompletedTaks: false,
        date: new Date().toString(),
      });
      setTask("");
      setModalVisible(false);
    } else {
      set(ref(db, "tarefas/" + oldTask), {
        id: oldTask,
        task: task,
        statusCompletedTask: false,
        date: new Date().toString(),
      });
      setTask("");
      setOldTask("");
      setModalVisible(false);
      setStatusAddTask(true);
    }
  };

  const completedTask = () => {
    set(ref(db, "tarefas/" + oldTask), {
      id: oldTask,
      task: task,
      statusCompletedTask: true,
      date: new Date().toString(),
    });
    setOldTask("");
    setModalVisible(false);
  };

  const deletedTask = () => {
    remove(ref(db, "tarefas/" + oldTask));
    setOldTask("");
    setModalVisible(false);
  };

  useEffect(() => {
    return onValue(ref(db, "/tarefas"), (querySnapShot) => {
      let data = querySnapShot.val() || {};
      let allItems = { ...data };
      setAllTasks(allItems);
    });
  }, []);

  const allTasksFireabase = Object.keys(allTasks);

  return (
    <View style={styles.container}>
      <View style={styles.containerTitle}>
        <Text style={styles.title} testID="title-list">
          Lista de Tarefas
        </Text>
      </View>
      <TouchableOpacity
        style={styles.buttonAdd}
        onPress={() => {
          setModalVisible(true);
          setTask("");
        }}
      >
        <Text style={styles.textAdd}>+</Text>
      </TouchableOpacity>

      <SafeAreaView style={styles.containerSafeArea}>
        <ScrollView style={styles.scrollView}>
          {allTasksFireabase.map((index) => (
            <View style={styles.containerScroll} key={allTasks[index]}>
              <Text
                style={styles.textScroll}
                key={allTasks[index]}
                onPress={() => {
                  if (!allTasks[index].statusCompletedTask) {
                    setModalVisible(true);
                  }
                  setTask(allTasks[index].task);
                  setStatusAddTask(false);
                  setOldTask(allTasks[index].id);
                }}
              >
                {allTasks[index].task} -{" "}
                {moment(allTasks[index].date).format("HH:mm")}
              </Text>
              <Text style={styles.iconCompleted}>
                {allTasks[index].statusCompletedTask ? (
                  <Text style={{ color: "green" }}>Check</Text>
                ) : null}
              </Text>
            </View>
          ))}
        </ScrollView>
      </SafeAreaView>

      <Modal
        transparent={true}
        visible={modalVisible}
        onRequestClose={() => {
          setModalVisible(!modalVisible);
        }}
      >
        <View style={styles.centeredView}>
          <View style={styles.modalView}>
            <TouchableOpacity
              style={styles.buttonClose}
              onPress={() => {
                setModalVisible(!modalVisible);
                setStatusAddTask(true);
              }}
            >
              <Text style={styles.textClose}>X</Text>
            </TouchableOpacity>

            <Text style={styles.titleModal}>
              {statusAddTask ? "Adicionar nova tarefa" : "Dados da Tarefa"}
            </Text>

            <View style={styles.containerForm}>
              <TextInput
                testID="input"
                style={styles.inputText}
                value={task}
                onChangeText={(event) => setTask(event)}
              />
              <TouchableOpacity style={styles.buttoAddTask} onPress={addTask}>
                <Text style={styles.textButtonAdd}>
                  {statusAddTask ? "Adicionar Tarefa" : "Atualizar Tarefa"}
                </Text>
              </TouchableOpacity>
              {!statusAddTask ? (
                <>
                  <TouchableOpacity
                    style={styles.buttonCompleted}
                    onPress={completedTask}
                  >
                    <Text style={styles.textButtonAdd}>Concluir</Text>
                  </TouchableOpacity>
                  <TouchableOpacity
                    style={styles.buttonDeleted}
                    onPress={deletedTask}
                  >
                    <Text style={styles.textButtonAdd}>Deletar</Text>
                  </TouchableOpacity>
                </>
              ) : null}
            </View>
          </View>
        </View>
      </Modal>
    </View>
  );
};

export default Home;
