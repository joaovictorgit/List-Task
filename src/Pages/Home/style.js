import { Dimensions, StatusBar, StyleSheet } from "react-native";

const styles = StyleSheet.create({
  container: {
    width: Dimensions.get("screen").width,
    height: Dimensions.get("screen").height,
    flex: 1,
  },
  containerTitle: {
    flex: 1,
    flexDirection: "column",
    alignItems: "center",
  },
  title: {
    fontWeight: "bold",
    fontSize: 16,
    marginTop: 35,
  },
  buttonAdd: {
    flex: 1,
    backgroundColor: "blue",
    width: 60,
    height: 60,
    justifyContent: "center",
    alignItems: "center",
    borderRadius: 50,
    position: "absolute",
    bottom: 10,
    right: 10,
  },
  centeredView: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    marginTop: 22,
  },
  containerScroll: {
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    width: 500,
    padding: 15,    
    borderBottomWidth: 1,
    borderBottomColor: 'rgb(162, 162, 162)'
  },
  iconCompleted: {
    position: "relative",
    bottom: -5,
    right: -100,
  },
  modalView: {
    margin: 20,
    backgroundColor: "white",
    borderRadius: 20,
    padding: 35,
    alignItems: "center",
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
  },
  button: {
    borderRadius: 20,
    padding: 10,
    elevation: 2,
  },
  buttonClose: {
    position: "absolute",
    top: 5,
    right: 10,
  },
  textStyle: {
    color: "white",
    fontWeight: "bold",
    textAlign: "center",
  },
  modalText: {
    marginBottom: 15,
    textAlign: "center",
  },
  containerForm: {
    justifyContent: "center",
    alignItems: "center"
  },
  inputText: {
    borderBottomWidth: 1,
    width: 250,
    marginTop: 25,
  },
  titleModal: {
    fontSize: 14,
    fontWeight: "500",
  },    
  buttoAddTask: {
    backgroundColor: "blue",
    justifyContent: "center",
    alignItems: "center",
    marginTop: 25,
    borderRadius: 5,
    width: 150,
    height: 30,
  },
  buttonCompleted: {
    backgroundColor: "rgba(0, 160, 0, 0.64)",
    justifyContent: "center",
    alignItems: "center",
    marginTop: 10,
    borderRadius: 5,
    width: 150,
    height: 30,
  },
  buttonDeleted: {
    backgroundColor: "rgba(178, 34, 34, 0.8)",
    justifyContent: "center",
    alignItems: "center",
    marginTop: 10,
    borderRadius: 5,
    width: 150,
    height: 30,
  },
  textButtonAdd: {
    color: "white"
  },
  containerSafeArea: {
    flex: 1,
    position: "absolute",
    left: 0,
    right: 0,
    top: 100,
  },
  scrollView: {
    marginHorizontal: 20
  },
  textScroll: {
    fontSize: 16,
    color: 'rgb(162, 162, 162)',
    fontWeight: "400",
    position: "absolute",
    left: 0,
    bottom: 10,
  }
});

export default styles;
