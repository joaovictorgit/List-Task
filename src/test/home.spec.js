import { render, screen, fireEvent } from "@testing-library/react-native";
import Home from "../Pages/Home/home-screen";

describe("Home Screen", () => {
  render(<Home />);
  it("should be able title home screen", () => {
    expect(screen.getByText("Lista de Tarefas"));
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
