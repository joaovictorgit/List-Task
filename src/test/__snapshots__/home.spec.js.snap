// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen should be able title home screen 1`] = `
<View
  style={
    {
      "flex": 1,
      "height": 1334,
      "width": 750,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 16,
          "fontWeight": "bold",
          "marginTop": 35,
        }
      }
      testID="title-list"
    >
      Lista de Tarefas
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "blue",
        "borderRadius": 50,
        "bottom": 10,
        "flex": 1,
        "height": 60,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 10,
        "width": 60,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 25,
        }
      }
    >
      +
    </Text>
  </View>
  <RCTSafeAreaView
    style={
      {
        "flex": 1,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 100,
      }
    }
  >
    <RCTScrollView
      style={
        {
          "marginHorizontal": 20,
        }
      }
    >
      <View />
    </RCTScrollView>
  </RCTSafeAreaView>
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
</View>
`;
